### PR TITLE
Add disablePrototypePoisoningProtection configuration

### PR DESCRIPTION
**Pull request by kavilla**
### Description
Enables the configuration of `disablePrototypePoisoningProtection` by setting
`opensearch.disablePrototypePoisoningProtection`. Enables users to store
protected logs that include reserve words from JS without the
OpenSearch JS client throwing errors.

We should still consider transforming unsafe data values if a bad actor
attempts to prototype pollute the cluster.

More information:
https://web.archive.org/web/20200319091159/https://hueniverse.com/square-brackets-are-the-enemy-ff5b9fd8a3e8?gi=184a27ee2a08

Related issue:
https://github.com/opensearch-project/OpenSearch-Dashboards/issues/1777

Signed-off-by: Kawika Avilla <kavilla414@gmail.com>
 
### Check List
- [ ] All tests pass
  - [ ] `yarn test:jest`
  - [ ] `yarn test:jest_integration`
  - [ ] `yarn test:ftr`
- [x] New functionality includes testing.
- [x] New functionality has been documented.
- [x] Update [CHANGELOG.md](./../CHANGELOG.md)
- [x] Commits are signed per the DCO using --signoff

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -46,6 +46,7 @@ Inspired from [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
 - [Vis Builder] Add field summary popovers ([#2682](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/2682))
 - [I18n] Register ru, ru-RU locale ([#2817](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/2817))
 - Add yarn opensearch arg to setup plugin dependencies ([#2544](https://github.com/opensearch-project/OpenSearch-Dashboards/issues/2544))
+- Add disablePrototypePoisoningProtection configuration ([#2992](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/2992))
```
Comment (AMoo-Miki) [inline CHANGELOG.md:49]: Can we make this a little more descriptive?

File: src/core/server/opensearch/client/client_config.ts [hunk 2]
```diff
@@ -115,6 +116,10 @@ export function parseClientOptions(config: OpenSearchClientConfig, scoped: boole
     );
   }
 
+  if (config.disablePrototypePoisoningProtection != null) {
+    clientOptions.disablePrototypePoisoningProtection = config.disablePrototypePoisoningProtection;
```
Comment (AMoo-Miki) [inline src/core/server/opensearch/client/client_config.ts:120]: How would this handle "false" as a string?

File: CHANGELOG.md (hunk 1 above)
Comment (kavilla) [reply CHANGELOG.md]: Update!

File: src/core/server/opensearch/client/client_config.ts (hunk 2 above)
Comment (kavilla) [reply src/core/server/opensearch/client/client_config.ts:120]: Good call out. Right now it's defined as optional or boolean so not able to pass in a string without failing to bootstrap.

**Review by AMoo-Miki — DISMISSED** [review]

**Comment by seraphjiang** [comment]
@kavilla 
Reading through code. didn't find the new config being used anywhere. do i missed or it is intended to just add a config item.

**Comment by kavilla** [comment]
> @kavilla Reading through code. didn't find the new config being used anywhere. do i missed or it is intended to just add a config item.

Yes it will be a new config to be used if any SaaS providers want to set `opensearch.disablePrototypePoisoningProtection: true` in their `opensearch_dashboards.yml`.

File: config/opensearch_dashboards.yml [hunk 3]
```diff
@@ -107,6 +107,9 @@
 # Logs queries sent to OpenSearch. Requires logging.verbose set to true.
 #opensearch.logQueries: false
 
+# Disables errors from client and enables you to utilize protected words within cluster
```
Comment (seraphjiang) [inline config/opensearch_dashboards.yml:110]: could we add a little information to share this config is passing to initialize opensearch-js

Comment (joshuarrrr) [reply config/opensearch_dashboards.yml:110]: Do we need any sort of warning about what the gotchas are of disabling this? I assume there's a good reason for why we don't want to enable this by default.

**Comment by joshuarrrr** [comment]
@kavilla Marking this as a draft until you have a chance to pick it back up.

**Comment by ratishr** [comment]
@kavilla  By when will this be reviewed and committed? We are seeing more and more such issues

File: config/opensearch_dashboards.yml (hunk 3 above)
Comment (kavilla) [reply config/opensearch_dashboards.yml]: Addressed @seraphjiang @joshuarrrr

**Comment by ananzh** [comment]
@kavilla I think this PR is to add another new config. But I don't quite understand how this new config will solve the issue...
my point is that I am expecting a logic that some core service calling `opensearch.disablePrototypePoisoningProtection` then use the setting to do the check. If the config is true, then we relieve the protection and allow to use constructor and other name types.. could you help me to understand more on this?

**Review by ananzh — APPROVED** [review]

**Review by abbyhu2000 — APPROVED** [review]

**Comment by ananzh** [comment]
Backport to 1.x and 1.3 fail due to disablePrototypePoisoningProtection?: boolean | 'proto' | 'constructor'; doesn’t exist in "@elastic/elasticsearch": "7.10.0-rc.1" which is what we use in 1.x.

Therefore, I will remove the v1.3.8 tag

**Pull request merged**